x/programs/rust/examples/nft/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+
+    // Store the original owner of the NFT (the contract creator)
+    program
+        .state()
+        .store(
+            StateKey::Owner(owner).to_vec(),
+            &Address::new([0; 32]).as_bytes(),
+        )
+        .expect("failed to store program owner");
+
+    // set total supply
+    program
+        .state()
+        .store(StateKey::TotalSupply.to_vec(), &NFT_TOTAL_SUPPLY)
+        .expect("failed to store total supply");
+
+    true
+}
+
+/// Mints NFT tokens and sends them to the recipient.
+#[public]
+pub fn mint(program: Program, recipient: Address, quantity: i64) -> bool {
+    assert!(quantity > 0, "quantity must be greater than zero");
+    assert!(
+        quantity <= NFT_TOTAL_SUPPLY as i64,
+        "quantity must be less than or equal to total supply"
+    );
+
+    let balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(recipient).to_vec())
+        .expect("failed to get balance");
+
+    program
+        .state()
+        .store(StateKey::Balance(recipient).to_vec(), &(balance + quantity))
+        .expect("failed to store balance");
+
+    true
+}
+
+/// Transfers balance from the sender to the the recipient.
+#[public]
+pub fn transfer(program: Program, sender: Address, recipient: Address, amount: i64) -> bool {
+    assert_eq!(sender, recipient, "sender and recipient must be different");
+
+    // ensure the sender has adequate balance
+    let sender_balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(sender).to_vec())
+        .expect("failed to update balance");
+
+    assert!(
+        amount >= 0 && sender_balance >= amount,
+        "sender and recipient must be different"
+    );
+
+    assert_eq!(sender, recipient, "sender and recipient must be different");
+
+    let recipient_balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(recipient).to_vec())
+        .expect("failed to store balance");
+
+    // update balances
+    program
+        .state()
+        .store(
+            StateKey::Balance(sender).to_vec(),
+            &(sender_balance - amount),
+        )
+        .expect("failed to store balance");
+
+    program
+        .state()
+        .store(
+            StateKey::Balance(recipient).to_vec(),
+            &(recipient_balance + amount),
+        )
+        .expect("failed to store balance");
+
+    true
+}
+
+#[public]
+pub fn burn(program: Program, from: Address, amount_burned: i64) -> bool {
+    assert!(amount_burned > 0, "quantity must be greater than zero");
+
+    let balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(from).to_vec())
+        .expect("failed to get balance");
+
+    assert!(
+        amount_burned <= balance,
+        "quantity must be less than or equal to total supply"
+    );
+
+    program
+        .state()
+        .store(StateKey::Balance(from).to_vec(), &(balance - amount_burned))
+        .expect("failed to store new balance");
```

Comment:
question: does burning also reduce the total supply?

Reply:
Yes, I believe it should. Good catch.